cmd/kafka-consumer/writer.go

```diff
@@ -146,391 +101,382 @@ func newWriter(ctx context.Context, o *option) *writer {
-				for _, row := range cachedEvents {
-					w.checkPartition(row, partition, message.TopicPartition.Offset)
-					log.Info("simple protocol cached event resolved, append to the group",
-						zap.Int64("tableID", row.GetTableID()), zap.Uint64("commitTs", row.CommitTs),
-						zap.Int32("partition", partition), zap.Any("offset", offset))
-					w.appendRow2Group(row, progress, offset)
-				}
-			}
+		// the Query maybe empty if using simple protocol, it's comes from `bootstrap` event, no need to handle it.
+		if ddl.Query == "" {
+			return false
+		}
 
-			// the Query maybe empty if using simple protocol, it's comes from `bootstrap` event, no need to handle it.
-			if ddl.Query == "" {
-				continue
-			}
+		log.Info("DDL event received",
+			zap.Int32("partition", partition), zap.Any("offset", offset),
+			zap.Uint64("commitTs", ddl.GetCommitTs()), zap.String("query", ddl.Query),
+			zap.Any("blockedTables", ddl.GetBlockedTables()), zap.Any("DDL", ddl))
 
-			if partition == 0 {
-				w.appendDDL(ddl, offset)
-			}
-			needFlush = true
-		case model.MessageTypeRow:
-			row, err := progress.decoder.NextRowChangedEvent()
-			if err != nil {
-				log.Panic("decode message value failed",
-					zap.Int32("partition", partition), zap.Any("offset", offset),
-					zap.ByteString("value", value),
-					zap.Error(err))
+		err := w.flushDDLEvent(ctx, ddl)
+		if err != nil {
+			log.Panic("flush DDL event failed", zap.String("DDL", ddl.Query), zap.Error(err))
+		}
+		return true
+	case common.MessageTypeRow:
+		var counter int
+		row := progress.decoder.NextDMLEvent()
+		if row == nil {
+			if w.protocol != config.ProtocolSimple {
+				log.Panic("DML event is nil, it's not expected",
+					zap.Int32("partition", partition), zap.Any("offset", offset))
 			}
-			// when using simple protocol, the row may be nil, since it's table info not received yet,
-			// it's cached in the decoder, so just continue here.
-			if w.option.protocol == config.ProtocolSimple && row == nil {
-				continue
+			log.Warn("DML event is nil, it's cached ", zap.Int32("partition", partition), zap.Any("offset", offset))
+			break
+		}
+
+		w.appendRow2Group(row, progress, offset)
+		counter++
+		for {
+			_, hasNext = progress.decoder.HasNext()
+			if !hasNext {
+				break
 			}
-			w.checkPartition(row, partition, message.TopicPartition.Offset)
+			row = progress.decoder.NextDMLEvent()
 			w.appendRow2Group(row, progress, offset)
-		case model.MessageTypeResolved:
-			newWatermark, err := progress.decoder.NextResolvedEvent()
-			if err != nil {
-				log.Panic("decode message value failed",
-					zap.Int32("partition", partition), zap.Any("offset", offset),
-					zap.ByteString("value", value), zap.Error(err))
-			}
-
-			progress.updateWatermark(newWatermark, offset)
-			w.resolveRowChangedEvents(progress, newWatermark)
-			needFlush = true
-		default:
-			log.Panic("unknown message type", zap.Any("messageType", messageType),
+			counter++
+		}
```

Comment:
Why does it check the next event here but not the first `NextDMLEvent`? It seems redundant.